commands/deploy_test.go

```diff
@@ -48,3 +55,51 @@ func TestCommandsDeployBuildFilelistErrorsOnNonExistentDirectory(t *testing.T) {
 		})
 	}
 }
+
+func TestCommandsDeployUploadsTarball(t *testing.T) {
+	assert := assert.New(t)
+
+	// Setup
+	var req struct {
+		called   bool
+		username string
+		password string
+		body     []byte
+	}
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		u, p, ok := r.BasicAuth()
+		assert.True(ok)
+		req.called = true
+		req.username = u
+		req.password = p
+		b, err := ioutil.ReadAll(r.Body)
+		assert.NoError(err)
+		req.body = b
+		w.WriteHeader(http.StatusOK)
+	}))
+
+	url, err := url.Parse(ts.URL)
+	assert.NoError(err)
+
+	auth.CredentialPath = newCredentialTempfile(t)
+	endpoint := url.Host
+	username := "hello"
+	password := "s3cr3t"
+	auth.WriteCredential(endpoint, username, password)
+
+	dir := filepath.Join("testdata", "deploy", "tree")
+
+	// Invoke
+	c := DeployCmd{Directory: dir, ServerURL: url}
+	err = c.Run()
+
+	assert.NoError(err)
+	assert.True(req.called)
+	assert.Equal(username, req.username)
+	assert.Equal(password, req.password)
+	assert.NotZero(len(req.body))
+
+	mime, err := mimetype.DetectReader(bytes.NewReader(req.body))
```

Comment:
You could just check if the first two bytes are 0x1f and 0x8b, i.e. the gzip header, and save the ongoing maintenance of bringing in another third party dependency. Check the third byte is 0x08 if deflate compression is important.

Reply:
That's quite a pro-tip! I'll change that now.